libvast/vast/address_synopsis.hpp

```diff
@@ -48,6 +54,66 @@ class address_synopsis final
   }
 };
 
+/// A synopsis for IP addresses that stores a copy of the input to
+/// be able to construct a smaller bloom filter from this data at
+/// some point using the `shrink` function.
+template <class HashFunction>
+class buffered_address_synopsis
+  : public bloom_filter_synopsis<address, HashFunction> {
+public:
+  using super = bloom_filter_synopsis<address, HashFunction>;
+
+  buffered_address_synopsis(type x, typename super::bloom_filter_type bf)
+    : super{std::move(x), std::move(bf)} {
+  }
+
+  void add(data_view x) override {
+    auto addr_view = caf::get_if<view<address>>(&x);
+    VAST_ASSERT(addr_view);
+    ips_.push_back(*addr_view);
+    super::add(x);
+  }
+
+  synopsis_ptr shrink() override {
+    // The bloom_filter doesnt store its `p`, so we parse the type to get it,
+    // which was enriched with this information in `make_address_synopsis()`.
+    auto& type = this->type();
+    auto old_params = parse_parameters(type);
+    if (!old_params)
+      return nullptr;
+    std::sort(ips_.begin(), ips_.end());
+    auto end = std::unique(ips_.begin(), ips_.end());
+    auto begin = ips_.begin();
+    auto n = std::distance(begin, end);
+    ips_.resize(n);
+    size_t next_power_of_two = 1ull;
+    while (ips_.size() > next_power_of_two)
+      next_power_of_two *= 2;
```

Comment:
So what you want is `msb(ips_.size()) + 1`?

Reply:
To be exact: `is_exact_power_of_two(ips_.size()) ? ips_.size() : 2*msb(ips_.size())`.

But this is now obsoleted with the latest revision.